grid-client/calculator/calculate.go

```diff
@@ -143,6 +152,314 @@ func calculateCU(cru, mru int64) float64 {
+	if err != nil {
+		return 0, err
+	}
+
+	node, err := c.substrateConn.GetNode(nodeID)
+	if err != nil {
+		return 0, err
+	}
+
+	var nodeRentContract uint64
+
+	nodeRentContract, err = c.substrateConn.GetNodeRentContract(nodeID)
+	if err != nil && !errors.Is(err, substrate.ErrNotFound) {
+		return 0, err
+	}
+
+	if contract.ContractType.IsNodeContract {
+		return c.calculateNodeContractCost(contract, node.Certification.IsCertified, nodeRentContract > 0)
+	}
+
+	if contract.ContractType.IsRentContract {
+
+		return c.calculateRentCost(contract, *node)
+	}
+	return 0, nil
+}
+
+// Calculates the cost of a unique name per month in USD.
+func (c *Calculator) calculateUniqueNameCost() (float64, error) {
+	//TODO should we apply stacking discount?
+	pricingPolicy, err := c.substrateConn.GetPricingPolicy(defaultPricingPolicyID)
+	if err != nil {
+		return 0, err
+	}
+	// cost in unit-USD
+	monthlyCost := float64(pricingPolicy.UniqueName.Value) * 24 * 30
+	return float64(monthlyCost) / UnitFactor, nil
+}
+
+// Calculates the cost of a node contract per month in USD.
+//
+// There are two cases for node contract cost:
+//  1. Node contract on shared node: the cost of the node (shared)
+//  2. Node contract on rented node: the cost of the IPV4 only if the contact includes ipv4, else it will return zero.
+func (c *Calculator) calculateNodeContractCost(contract *substrate.Contract, onCertifiedNode, isOnRentedNode bool) (float64, error) {
+	if !contract.ContractType.IsNodeContract {
+		return 0, fmt.Errorf("contract ID %d is not a node contract", contract.ContractID)
+	}
+	publicIPsCount := contract.ContractType.NodeContract.PublicIPsCount
+
+	isCertified := node.Certification.IsCertified
+
+	// Node contract on rented node
+	if isOnRentedNode {
+		if publicIPsCount > 0 {
+			cost, err := c.calculateIPV4()
+			if err != nil {
+				return 0, err
+			}
+			totalCost := cost * float64(publicIPsCount)
+
+			//TODO should we apply stacking discount?
+
+			if onCertifiedNode {
+				totalCost *= 1.25
+			}
+			return totalCost, nil
+		}
+		return 0, nil
+	}
+
+	// Normal node contract on sharedNode
+
+	resources, err := c.substrateConn.GetNodeContractResources(uint64(contract.ContractID))
+	if err != nil {
+		return 0, err
+	}
+	CRU := resources.Used.CRU
+	MRU := convertBytesToGB(uint64(resources.Used.MRU))
+	HRU := convertBytesToGB(uint64(resources.Used.HRU))
+	SRU := convertBytesToGB(uint64(resources.Used.SRU))
+
+	cost, err := c.CalculateCost(int64(CRU), int64(MRU), int64(HRU), int64(SRU), publicIPsCount > 0, onCertifiedNode)
+	if err != nil {
+		return 0, err
+	}
+	_, sharedPrice, err := c.CalculatePricesAfterDiscount(cost)
+	if err != nil {
+		return 0, err
+	}
+	return sharedPrice, nil
+}
+
+// Calculates the cost of a rent contract per month in USD.
+//
+// Rent contract cost is the cost of the node (dedicated discount applied) + the node extra fee
+func (c *Calculator) calculateRentCost(contract *substrate.Contract, node substrate.Node) (float64, error) {
```

Comment:
why we don't calculate ips?

Reply:
the ips is assosiated with node contracts only